concurrency/MpmcPublish.h

```diff
@@ -14,82 +16,77 @@ class MpmcPublish
     {
         Entry() : refCount(0) {}
 
-        std::atomic_uint32_t refCount;
+        std::atomic_uint32_t refCount; // 0 means free
         T value;
     };
 
-    static constexpr uint32_t indexBits = 16;
-    static constexpr uint32_t TOO_MANY_THREADS = 1 << (indexBits - 1);
-    static constexpr uint32_t indexMask = (1 << indexBits) - 1;
-
-    uint32_t indexOf(uint32_t cursor) const { return cursor & indexMask; }
-    uint32_t versionOf(uint32_t cursor) const { return cursor >> indexBits; }
-    uint32_t toCursor(uint32_t version, uint32_t index) const { return (version << indexBits) | 0x80000000 | index; }
-
 public:
     typedef T ValueType;
-    MpmcPublish() : _readCursor(0)
+    MpmcPublish() : _readPointer(nullptr)
     {
-        static_assert(MAX_THREADS > 0 && MAX_THREADS < TOO_MANY_THREADS,
-            "entries in MpmcPublish cannot exceed TOO_MANY_THREADS");
+        static_assert(MAX_THREADS > 2 && MAX_THREADS < 512, "entries must > 2 < 512 in MpmcPublish");
     }
 
     ~MpmcPublish() {}
 
     // return false if empty
     bool read(T& target) const
     {
-        if (empty())
-        {
-            return false;
-        }
-
-        for (;;)
+        for (auto* cell = _readPointer.load(std::memory_order_consume); cell;
+             cell = _readPointer.load(std::memory_order_consume))
         {
-            const uint32_t readIndex = indexOf(_readCursor.load(std::memory_order::memory_order_consume));
-            Entry& cell = const_cast<Entry&>(_elements[readIndex]);
-            if (cell.refCount.fetch_add(1) < TOO_MANY_THREADS)
+            const auto prevCount = cell->refCount.fetch_add(1);
+            if (prevCount > 0)
             {
-                target = cell.value;
-                cell.refCount.fetch_sub(1);
+                target = cell->value;
+                cell->refCount.fetch_sub(1);
                 return true;
             }
-            else
-            {
-                cell.refCount.fetch_sub(1);
-            }
+
+            cell->refCount.fetch_sub(1);
         }
+
+        return false;
     }
 
     // iterate until we find an idle slot to write into.
     // most reading threads will be locking the previous write pos
     bool write(const T& obj)
     {
-        uint32_t readCursor = _readCursor.load(std::memory_order::memory_order_relaxed);
-        const uint32_t readIndex = indexOf(readCursor);
-        const uint32_t version = versionOf(readCursor);
-        for (int i = 0; i < MAX_THREADS - 1; ++i)
+        auto readPointer = _readPointer.load(std::memory_order_consume);
```

Comment:
Simpler and more efficient write. In most likely case there are  load + fetchadd + cmpexch.
Before it had to: load, cmpexch, fetchsub, cmpexch, fetchsub